src/RamTransforms.cpp

```diff
@@ -641,84 +641,95 @@ bool TupleIdTransformer::reorderOperations(RamProgram& program) {
 bool HoistAggregateTransformer::hoistAggregate(RamProgram& program) {
     bool changed = false;
 
-    // hoist a single aggregate to an outer scope
-    // most outer permissible aggregate is hoisted
+    // There are two cases: aggregates that have no data-dependencies on
+    // other RAM operations, and aggregates that have data-dependencies.
+
+    // Hoist a single aggregate to an outer scope that
+    // has no data-dependency.
+    visitDepthFirst(program, [&](const RamQuery& query) {
+        std::unique_ptr<RamNestedOperation> newAgg;
+        int lastTupleOp = -1;
+
+        // The rewriter has two tasks: (1) identify a single aggregate that
+        // can be hoisted and (2) insert it at the outermost level.
+        // We assume all RamOperations are renumbered for this transformation.
```

Comment:
Both rewriters have the same description, so they could perhaps be more narrowly descriptive. I don't understand the point of the comment about renumbering of RamOperations, so either the comment needs changing or I need to learn more - not sure which.